### ISSUE TITLE
[TF 2.0] CrossShardOptimizer not working with global_step parameter

### ISSUE THREAD

**Issue by captain-pool** [tensorflow/tensorflow#29446]
<em>Please make sure that this is a bug. As per our [GitHub Policy](https://github.com/tensorflow/tensorflow/blob/master/ISSUES.md), we only address code/doc bugs, performance issues, feature requests and build/installation issues on GitHub. tag:bug_template</em>

**System information**
- Have I written custom code (as opposed to using a stock example script provided in TensorFlow): Ubuntu 14.04
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04):
- Mobile device (e.g. iPhone 8, Pixel 2, Samsung Galaxy) if the issue happens on mobile device:
- TensorFlow installed from (source or binary): Binary
- TensorFlow version (use command below): tensorflow==2.0.0.a
- Python version: 3.6
- Bazel version (if compiling from source):
- GCC/Compiler version (if compiling from source):
- CUDA/cuDNN version:
- GPU model and memory:

You can collect some of this information using our environment capture
[script](https://github.com/tensorflow/tensorflow/tree/master/tools/tf_env_collect.sh)
You can also obtain the TensorFlow version with: 1. TF 1.0: `python -c "import
tensorflow as tf; print(tf.GIT_VERSION, tf.VERSION)"` 2. TF 2.0: `python -c
"import tensorflow as tf; print(tf.version.GIT_VERSION, tf.version.VERSION)"`

**Describe the current behavior**
I'm trying to run CrossShardOptimizer with tf.optimizers.Adam, however, tf.Optimizer.Adam is updated to v2 (with no global_step_parameter), and CrossShardOptimizer is still in version 1 and still have that global_step parameter. So, CrossShardOptimizer is failing Badly.
**Describe the expected behavior**
The CrossShardOptimizer shouldn't break.
**Code to reproduce the issue**
Provide a reproducible test case that is the bare minimum necessary to generate the problem.
```python
def model_fn(features, labels, mode, params):
    model = tf.keras.Sequential([
        tf.keras.layers.Flatten(input_shape=[28, 28, 1]),
        tf.keras.layers.Dense(128),
        tf.keras.layers.Dense(10, activation="softmax")
    ])
    optimizer = None
    if mode == tf.estimator.ModeKeys.TRAIN:
        optimizer = tf.optimizers.Adam(params.get("learing_rate", 1e-3))
        if params.get("use_tpu", True):
          optimizer = tpu_optimizer.CrossShardOptimizer(optimizer)

    with tf.GradientTape() as tape:
        logits = model(features)
        if mode == tf.estimator.ModeKeys.PREDICT:
            preds = {
                "predictions": logits
            }
            return tpu_estimator.TPUEstimatorSpec(mode, predictions=preds)
        loss = tf.keras.losses.SparseCategoricalCrossentropy(
            from_logits=True)(labels, logits)
    if mode == tf.estimator.ModeKeys.EVAL:
        return tpu_estimator.TPUEstimatorSpec(mode, loss=loss)

    def train_fn():
        assert optimizer is not None
        gradient = tape.gradient(loss, model.trainable_variables)
        global_step = tf.compat.v1.train.get_global_step()
        update_global_step = tf.compat.v1.assign(global_step, global_step + 1, name='update_global_step')
        with tf.control_dependencies([update_global_step]):
          apply_grads = optimizer.apply_gradients(zip(gradient, model.trainable_variables))
        return apply_grads

    if mode == tf.estimator.ModeKeys.TRAIN:
return tpu_estimator.TPUEstimatorSpec(mode, loss=loss, train_op=train_fn())
```
The complete code can be found at: 
https://github.com/captain-pool/tf2.0-tpu-sample/blob/master/image_retraining_tpu.py
**Other info / logs**
This issue gets fixed when:
https://github.com/tensorflow/tensorflow/blob/eed4d9c4c4f42c1c5338aa0bdca34c6efb956577/tensorflow/python/tpu/tpu_optimizer.py#L173
line is modified to
`self._opt.apply_gradients(summed_grads_and_vars, name=name)`
and updating `global_step` manually.
Like this:
https://github.com/captain-pool/tf2.0-tpu-sample/blob/39d2af5caf1dfbb28d0821b5143a9c534aef6861/image_retraining_tpu.py#L75-L78

**Do you want to contribute?**
Yes.

**Comment by captain-pool** [comment]
CC: @vbardiovskyg  @srjoglekar246 @dynamicwebpaige

**Comment by gadagashwini** [comment]
@captain-pool Looks like Complete code link provided above is pointing to error 404.Can you please check and point to the correct link. Thanks!

**Comment by captain-pool** [comment]
@gadagashwini Done :+1:

**Comment by jhseu** [comment]
I'm not sure it would be a good idea to support this because of the changes in optimizer would make its behavior confusing. It definitely could use a better error message, though, which I'll add.

You can do this instead to sum the gradients across replicas:
```
gradient = [tf.compat.v1.tpu.cross_replica_sum(grad) for grad in gradient]
apply_grads = optimizer.apply_gradients(zip(gradient, model.trainable_variables))
```

If you want to average the gradients, you'll need to also scale the loss down (divide by number of replicas).

The reason it would be confusing with OptimizerV2 is:
- OptimizerV2 doesn't have compute_gradients, where we rescale the loss.
- OptimizerV2 when used with TPUStrategy will already aggregate the gradients across replicas for you, and so it may end up being done twice.

**Comment by jhseu** [comment]
https://github.com/tensorflow/tensorflow/commit/06349f1973ccdd25c9d12ed495c519ebf6f3d797

We throw an exception now with the suggested workaround.
